xdsl/dialects/pdl_interp.py

```diff
@@ -417,6 +422,48 @@ def __init__(
         super().__init__(operands=[lhs, rhs], successors=[trueDest, falseDest])
 
 
+@irdl_op_definition
+class ApplyConstraintOp(IRDLOperation):
+    """
+    See external [documentation](https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpapply_constraint-pdl_interpapplyconstraintop).
+    """
+
+    name = "pdl_interp.apply_constraint"
+    traits = traits_def(IsTerminator())
+    constraint_name = prop_def(StringAttr, prop_name="name")
+    is_negated = prop_def(
+        BoolAttr, prop_name="isNegated", default_value=BoolAttr(False, i1)
+    )
+    args = var_operand_def(AnyPDLTypeConstr)
+    rresults = var_result_def(AnyPDLTypeConstr)
+    true_dest = successor_def()
+    false_dest = successor_def()
```

Comment:
Would it be worth sticking to the upstream names of `match_dest` and `failure_dest`?

Reply:
ah, but upstream is `trueDest` and `falseDest`, they just don't use those names in the example code.